### ISSUE TITLE
implement special handling of special statements

### ISSUE THREAD

**Issue by ivg** [BinaryAnalysisPlatform/bap#852]
So far when we translate from BIL to BIR we ignore special statements. An alternative approach would be to represent each special as a call to a synthetic subroutine. The main question is shall we redirect all to the same synthetic "undefined" or generate for each special its own subroutine. The latter would be helpful, as we then can model/stub them with a Primus Lisp, e.g., `special("sqrt")`. 

However, the problem is that `Special` is parametrized with an arbitrary string, which is usually more like a comment. The proposed solution is the following. We will define a grammar for the statements, that will still recognize all possible strings, but we be able to distinguish between named and unnamed special statements. All unnamed special statements will be represented as calls to a special "__bap_undefined" intrinsic function, while the rest will get their names as ascribed by the user. 

```
special-annotation ::= <named> | <unnamed>
named ::= @<ident> "<text>
ident ::= ([:ascii:] - [:whitespace:]')+
unnamed ::= <text>
text ::= <char>*
```

I.e., an identifier is any nonempty sequence of ascii characters without whitespaces. The named special statement is either an arbitrary text, or an identifier prefixed with the `@` character, e.g.,
```
@float-sine - computes float sine function
@x86-lfence
undefined behavior
``` 

The first is a function with an optional comment, the second is just a function name, the latter is an unnamed special. 

Note, this is a workaround for preserving interface compatibility in BAP 2.0 we will have advanced special statements that will allow a lifter writer to define uninterpreted functions including their interface and other properties. The proposed change will enable interoperation between BAP 1.0 and BAP 2.0 representations.

**Comment by dbrumley** [comment]
This seems reasonable, but also will introduce an inter-procedural flavor to all analysis where there may be no interprocedural flow.  It seems if you could add the ability to inline a user-defined (with a pre-defined nop) definition in a pass/option that would make life easier.

**Comment by ivg** [comment]
I agree, a pass that will ignore/inline a specified function would be damn useful, and will definitely alleviate the issue that you've mentioned. This change won't have ramifications right after introduction as we use specials very sparingly, if at all. But, yes, in the future, when we will start to use specials to represent unlifted instructions it will introduce that flavor.

**Comment by ivg** [comment]
a lighter solution that represents instructions that are unrepresentable in BIL as empty blocks is [implemented][1] in the `new-semantics` branch and will be merged after 1.6 is released

[1]: https://github.com/ivg/bap/commit/7c0d5c4ccf73f6d457d4cf38526c8569e6e8ff3d
